### PR TITLE
refactor: Adapt processing expired storage request to fisherman node based deletion process

### PR DESCRIPTION
**Pull request by HermanObst**
## Summary
Amend file deletion process for **expired** storage request. Followings PRs will handle revoked storage request and rejected (by MSP) storage requests.

### Changes
  - Implement `delete_file_for_incomplete_storage_request` extrinsic for cleaning up files from expired
   storage requests
  - Adapt expired storage request processing to new file deletion process to push an `IncompleteStorageRequestMetadata` to the new `IncompleteStorageRequests` mapping when there are providers to delete file from.
  - Add comprehensive test coverage for storage request expiration and BSP cleanup scenarios
  - Fix BSP removal logic during storage request cleanup

###  New types:
  - New type: `IncompleteStorageRequestMetadata` 
  - New event: `FileDeletedFromIncompleteStorageRequest`
  - New errors: `IncompleteStorageRequestNotFound`, `FileKeyMismatch`

###  Test Coverage:
  - Added failure cases for file deletion triggered by expired storage requests
  - Added tests for BSP deletion scenarios (single BSP with multiple files, multiple BSPs confirming then
  being deleted)

### Fixes:
  - Fixes `delete_file` extrinsic edge cases and a vulnerability where the extrinsic was not checking that the msp was indeed storing the given bucket, thus letting an attacker to delete the file_key from the msp storing it (and reducing the payment stream), but reducing the capacity of other msp the attacker wanted.

### CODE REVIEW

File: pallets/file-system/src/utils.rs [hunk 1]
```diff
@@ -3000,114 +3106,70 @@ mod hooks {
...
+                                ),
+                            );
+                        }
                         // Emit the StorageRequestRejected event
+                        // If there are BSPs that have confirmed storing the file,
+                        // this event will be used by the fisherman node to delete the file from the confirmed BSPs.
+                        // If there are no BSPs the event is just informative.
                         Self::deposit_event(Event::StorageRequestRejected {
                             file_key,
                             reason: RejectedStorageRequestReason::RequestExpired,
```
[the remark this answers is not in the document]
Comment (ffarall) [reply pallets/file-system/src/utils.rs:3171]: In that case, likely the fisherman task will see that there are no BSPs to delete and just do nothing.

File: pallets/file-system/src/utils.rs [hunk 2]
```diff
@@ -3000,114 +3106,70 @@ mod hooks {
...
-                            );
-                        });
-
-                        // Remove the storage request from the active storage requests for the bucket
-                        <BucketsWithStorageRequests<T>>::remove(
-                            &storage_request_metadata.bucket_id,
-                            &file_key,
-                        );
+                        // Clean up storage request data
+                        Self::cleanup_expired_storage_request(&file_key, &storage_request_metadata);
```
[the remark this answers is not in the document]
Comment (HermanObst) [reply pallets/file-system/src/utils.rs:3124]: If the MSP accepted the storage request and got "expired" as the target bsp cound didnt get to the required number, we anyways count this SR as accepted. 
So no more actions need to be taken.

File: pallets/file-system/src/utils.rs [hunk 3]
```diff
@@ -1300,6 +1302,110 @@ where
...
+        forest_proof: ForestProof<T>,
+    ) -> DispatchResult {
+        // Fetch storage request metadata
+        let mut storage_request_metadata =
+            StorageRequests::<T>::get(&file_key).ok_or(Error::<T>::StorageRequestNotFound)?;
+
+        // Ensure the storage request is rejected
+        ensure!(
+            storage_request_metadata.rejected,
+            Error::<T>::StorageRequestNotRejected
```
[the remark this answers is not in the document]
Comment (HermanObst) [reply pallets/file-system/src/utils.rs:1317]: Changed to check is the IncompleteStorageRequestMetadata exists

File: pallets/file-system/src/utils.rs [hunk 4]
```diff
@@ -1300,6 +1302,110 @@ where
...
+                storage_request_metadata.owner.clone(),
+                file_key,
+                storage_request_metadata.size,
+                storage_request_metadata.bucket_id,
+                provider_id,
+                forest_proof,
+            )?;
+
+            // Remove the MSP from the storage request after successful deletion
+            storage_request_metadata.msp = Some((provider_id, false));
```
[the remark this answers is not in the document]
Comment (HermanObst) [reply pallets/file-system/src/utils.rs:1354]: This is also fixed with changing to the new mapping approach, as all the StorageRequest related data will be removed upfront (as before), and the deletion flow will be handled by the new mapping.

File: pallets/file-system/src/utils.rs [hunk 5]
```diff
@@ -2662,6 +2729,15 @@ where
...
         forest_proof: ForestProof<T>,
     ) -> DispatchResult {
+        // Ensure that the provider_id is the owner of the bucket
+        ensure!(
+            <T::Providers as ReadBucketsInterface>::is_bucket_stored_by_msp(
+                &provider_id,
+                &bucket_id
+            ),
+            Error::<T>::MspNotStoringBucket
+        );
```
Comment (ffarall) [inline pallets/file-system/src/utils.rs:2739]: Good catch!!

File: pallets/file-system/src/utils.rs [hunk 6]
```diff
@@ -2821,21 +2897,54 @@ where
 
         Ok(())
     }
+
+    /// Construct IncompleteStorageRequestMetadata from existing storage request data
+    pub(crate) fn create_incomplete_storage_request_metadata(
+        storage_request: &StorageRequestMetadata<T>,
+        file_key: &MerkleHash<T>,
+    ) -> IncompleteStorageRequestMetadata<T> {
```
Comment (ffarall) [inline pallets/file-system/src/utils.rs:2905]: We could have implemented `Into<IncompleteStorageRequestMetadata> for &StorageRequestMetadata` or something like that. But this is perfectly fine too.

**Review by snowmead — APPROVED** [review]

File: pallets/file-system/src/tests.rs [hunk 7]
```diff
@@ -5861,7 +5861,7 @@ mod bsp_confirm {
 						// The deposit paid should have been updated after paying the BSP that volunteered.
 						deposit_paid: storage_request_deposit - <<Test as crate::Config>::WeightToFee as sp_weights::WeightToFee>::weight_to_fee(
 							&<Test as crate::Config>::WeightInfo::bsp_volunteer(),
-								),
+								)
```
[the remark this answers is not in the document]
Comment (HermanObst) [reply pallets/file-system/src/tests.rs:5864]: Before the `IncompleteStorageRequest` mapping refactor, I had added two new files to `StorageRequestMetadata`, which meant I had to include them in every test.
The missing comma issue happened when I deleted them all using find and replace.

If it looks too messy, I can fix it.
